### ISSUE TITLE
Develop a proof-of-concept prototype for a web-based graphical interface

### ISSUE THREAD

**Issue by niksirbi** [neuroinformatics-unit/movement#164]
## Feature idea: interactive visualisations in Jupyter
During [discussions](https://neuroinformatics.zulipchat.com/#narrow/stream/406001-Movement/topic/visualize.20tracking.20results) with @Di-Wang-AIND and Sharmishtaa Seshamani from the [Allen Institute for Neural Dynamics](https://alleninstitute.org/division/neural-dynamics/) we determined that there is a lot of interest in a web-based graphical interface form `movement`, specifically one that would allow for interactive data visualisations within Jupyter Notebooks. This could be very useful, since lots of researchers are comfortable with working in notebooks, especially during data exploration stages (where interactive visualisation is most needed). It would also facilitate cloud-based workflows.

## Context: web-based vs desktop (napari) frontends
For context, we are in parallel developing a desktop graphical interface via a napari plugin (see this [meta-issue](https://github.com/neuroinformatics-unit/movement/issues/31) and ongoing [work-in-progress](https://github.com/neuroinformatics-unit/movement/pull/112)). At this early stage, we think it makes sense to explore both options (napari-based and web-based) and evaluate their respective strengths and limitations. They should be independently developed but strive to present a consistent interface (as far as possible). Long-term we may decided to stay with only one frontend, or keep both.

## Example interactive visualisation workflow

This is an envisioned hypothetical workflow through a graphical frontend from the user's perspective:

- Select file containing pose predictions (probably via a file browser widget), the pose estimation software it was created with and the fps
- Optionally select the corresponding video file to show as underlay (could be another file browser widget). There should also be the option of using a static image as an underlay (may be a still frame or a schematic of the arena, for example)
- Explore the loaded data - the keypoints overlaid on the video/frame - with interactive widgets to adjust the shown individuals, keypoints, etc. (including option to plot a "tail" for a certain time window back in time).
- Generate and save video/clips or plots (for figures, presentations, reports). This part is less fleshed out in my mind, because so far I've been focusing on implementing the previous bits.

The end-product may look something like the mockup below:

![movement-gui-mockup_20240307](https://github.com/neuroinformatics-unit/movement/assets/20923448/4f4cd9b5-ed83-410f-ba6c-fb4d6a4f3c9b)

The above mockup is not to be taken literally, it's still on a conceptual level. The only part of it that's underway on the napari side is the file-loader widget:

<img width="1439" alt="Screenshot-2024-03-07-at-15 08 17" src="https://github.com/neuroinformatics-unit/movement/assets/20923448/97521e2f-806f-41df-a980-387f8678cd13">

## Potential frameworks to use
So far we've brought up [fastplotlib](https://fastplotlib.readthedocs.io/en/latest/) and [plotly](https://dash.plotly.com/).
That said, @Di-Wang-AIND has identified some potential blockers on the fastplotlib side, see https://github.com/fastplotlib/fastplotlib/issues/376 and https://github.com/fastplotlib/fastplotlib/issues/398. Alternative frameworks may still be considered.

## How to move forward
I think it would be best for this exploration to start on a separate branch, ideally leading to some demos. The first implementation need not be "clean", it should rather serve as a proof-of-concept. After evaluating that, we could move onto slowly merging the functionality into the main branch, ideally in the form of atomic PRs (each complete with tests).

@Di-Wang-AIND do let me know if you want to take this on within the next two months.

**Comment by niksirbi** [comment]
I've had a shot at starting en example notebook in [this branch](https://github.com/neuroinformatics-unit/movement/tree/fastplotlib-example). It's extremely bare bones for now:
- I use the `ImageWidget` for playing the video
- I overlayed animal trajectories over it (using a "line collection" graphic).
But the two elements are not synced yet; ideally in each video frame we should show the trajectory only for the last `n` frames (a "tail").

**Comment by kushalkolar** [comment]
You'll want to create your own widget from scratch if you want to use
pynapple and sync the time, Caitlin has a large example here:
https://github.com/fastplotlib/fastplotlib-fens2024/tree/main

To only make a part of the line visible you can start with adding a line
collection but pass it data for the first n_datapoints, and a have a
callback function that sets new data every time the time point changes.
Delete the existing line collection graphic if the size of the data changes
(gpu buffer sizes are fixed)

**Comment by niksirbi** [comment]
Thanks @kushalkolar, this us useful!

**Comment by kushalkolar** [comment]
Here's a quick example I wrote, not pynapple but get the basic fpl concepts across:

```python
import numpy as np
import fastplotlib as fpl
from ipywidgets import VBox

# generate some data

n_timepoints = 1_000

# txy
behavior_video = np.random.rand(n_timepoints, 512, 512)

# tracks
n_body_points = 5
behavior_tracks = np.zeros((n_body_points, n_timepoints, 2))  # n_body_points, n_timepoints, xy

# generate some tracks, start from center
behavior_tracks[:, 0] = np.array([256, 256])
for i in range(1, n_timepoints):
    behavior_tracks[0, i, 0] = behavior_tracks[0, i - 1, 0] + np.random.normal(scale=15)
    behavior_tracks[0, i, 1] = behavior_tracks[0, i - 1, 1] + np.random.normal(scale=15)

    for p in range(1, n_body_points):
        # just shift the other body points by a bit
        behavior_tracks[p, i] = behavior_tracks[0, i] + (p * 4)

    behavior_tracks = behavior_tracks.clip(0, 512)

# some other timeseries data, ephys, behavior
other_timeseries = np.random.rand(10, n_timepoints) * 10

# make the visualization

# number of tracks timepoints visible at one time
N_TIMEPOINTS_VISIBLE = 25

fig_behavior = fpl.Figure()
subplot_behavior = fig_behavior[0, 0]

# initialize with first frame of video
subplot_behavior.add_image(behavior_video[0], name="movie", cmap="grey")
subplot_behavior.add_line_collection(behavior_tracks[:, :N_TIMEPOINTS_VISIBLE], name="tracks", thickness=4)

# set color according to early and late time
subplot_behavior["tracks"].cmap[:] = "viridis"


fig2 = fpl.Figure()
fig2[0, 0].add_line_stack(other_timeseries, name="other-time-series")

# add a linear region selector
linear_selector = fig2[0, 0]["other-time-series"].add_linear_selector()


@linear_selector.add_event_handler("selection")
def update_time(event):
    ix = int(event.info["value"])
    if ix + N_TIMEPOINTS_VISIBLE > n_timepoints:
        return

    # set video frame
    subplot_behavior["movie"].data = behavior_video[ix]

    # set new xy data, note that line data are always 3d so select only xy dims
    for line, new_data in zip(fig_behavior[0, 0]["tracks"], behavior_tracks[:, ix:ix + N_TIMEPOINTS_VISIBLE]):
        line.data[:, :-1] = new_data


VBox([fig_behavior.show(), fig2.show(maintain_aspect=False)])
```

https://github.com/user-attachments/assets/a695f2cf-1a8e-414c-bd5a-c446c00eddb3

**Comment by niksirbi** [comment]
Looks amazing! This should enable to at least publish a "How to visualise movemen" data with fastplotlib" example in our docs.

**Comment by kushalkolar** [comment]
Btw I'm guessing your goal is to create a GUI? Note that we will be integrating imgui into fastplotlib very soon, we have been prototyping and plan to have a first version of imgui integration by the end of August. I have an early prototype in this messy PR: https://github.com/fastplotlib/fastplotlib/pull/552

Imgui is pretty easy to use, the nice thing is that it allows a uniform UI across all canvas backends, which means the same UI for jupyter and desktop use (qt or glfw), so there aren't separate modules to maintain for each.

You're of course welcome to use Qt or ipywidgets even after we integrate imgui.

**Comment by niksirbi** [comment]
Indeed, eventually we do want a GUI, but I thought I'd start with just an example figure, as our foray into exploring `fastplotlib`.

I'm not familiar with Imgui but what you say sounds very promising. I'll keep playing with `fastplotlib` and keep an eye on that PR and other relevant developments. We want to develop a feel for `fastplotlib` capabilities, to decide if it's the right tool for us. Of course, we're happy to keep you posted with feedback during that process.

**Comment by kushalkolar** [comment]
Just keeping you in the loop of where we're at with things relevant to your use case. Fastplotlib was made exactly for these types of visualizations, feel free to ask for help, simple examples like above only take me a few mins to make :)
